python/src/aiconfig/callback.py

```diff
@@ -0,0 +1,139 @@
+# Standard Libraries
+import asyncio
+import json
+import time
+
+# Third Party Libraries
+from result import Err, Ok
+from pydantic import BaseModel, ConfigDict
+from typing import (
+    Coroutine,
+    Final,
+    List,
+    Sequence,
+    TextIO,
+    TypeAlias,
+    Union,
+    Any,
+    Awaitable,
+    Callable,
+    Optional,
+)
+
+# Constants
+DEFAULT_TIMEOUT = 5  # Default timeout for callback execution in seconds
+
+
+class CallbackEvent:
+    """
+    Represents an event with data to be passed to a callback.
+    """
+
+    def __init__(self, name: str, data: Any, ts_ns: int = time.time_ns()):
+        self.name = name
+        # Anything available at the time the event happens.
+        # It is passed to the callback.
+        self.data = data
+        self.ts_ns = ts_ns
+
+
+# Type Aliases
+Callback = Callable[[CallbackEvent], Awaitable]
+Result: TypeAlias = Union[Ok[Any], Err[Any]]
+
+
+async def execute_coroutine_with_timeout(
+    coroutine: Coroutine[Any, Any, Any], timeout: int
+) -> Result:
+    """
+    Execute a coroutine with a timeout, return an Ok result or an Err on Exception
+
+    Args:
+        coroutine: The coroutine to execute
+        timeout: The timeout in seconds
+    
+    Returns:
+        An Ok result if the coroutine executes successfully or an Err on Exception
+    """
+    try:
+        task = asyncio.create_task(coroutine)
+        res = await asyncio.wait_for(task, timeout=timeout)
+        return Ok(res)
+    except Exception as e:
+        return Err(str(e))
+
+
+class Record(BaseModel):
+    model_config = ConfigDict(strict=True, frozen=True)
+
+
+class pydanticBackedCallBackEvent(Record):
```

Comment:
It can be a pydantic class directly, reason I chose to do it this way is so that you can construct the class easily.

`CallbackEvent("event", {"input": data})`

where as with Pydantic, you have to do dictionary unpacking
`CallbackEvent(**{"event", {"input": data}})`

There isn't a need for the type validation Pydantic brings here, just the easy serialization & validation of objects.